### PR TITLE
add where op with sparse condition

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
support sparse.where with condition being csr ndarray
#9464 need to be merged first
@reminisce 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/tensor/control_flow_op.h [hunk 1]
```diff
@@ -51,6 +52,35 @@ struct where {
...
+struct where_csr {
+  // DType is the output data type
+  // CType is condition data type
+  // i is for i-th row in the output
+  template<typename DType, typename CType, typename IType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, const IType* cond_idx,
+                                  const IType* cond_indptr, const CType* cond_data,
+                                  const nnvm::dim_t num_cols, const DType* x) {
+    using nnvm::dim_t;
+    dim_t offset = i * num_cols;
```
Comment (reminisce) [inline src/operator/tensor/control_flow_op.h]: const

File: src/operator/tensor/control_flow_op.h [hunk 2]
```diff
@@ -51,6 +52,35 @@ struct where {
...
+    for (dim_t j = cond_indptr[i]; j < cond_indptr[i + 1]; j++) {
+      const CType data = cond_data[j];
+      if (data != 0) {
+        const IType col_idx = cond_idx[j];
+        dim_t out_idx = offset + col_idx;
```
Comment (reminisce) [inline src/operator/tensor/control_flow_op.h]: const

File: src/operator/tensor/control_flow_op.h [hunk 3]
```diff
@@ -87,6 +117,37 @@ struct where_backward {
...
+                                  const DType* grad_in,
+                                  const CType* cond_data,
+                                  const IType* cond_idx,
+                                  const IType* cond_indptr,
+                                  const nnvm::dim_t num_cols) {
+    const IType offset = i * num_cols;
+    const DType zero = static_cast<DType>(0);
+    for (IType j = cond_indptr[i]; j < cond_indptr[i + 1]; j++) {
+      IType col = cond_idx[j];
+      IType grad_offset = offset + col;
```
Comment (reminisce) [inline src/operator/tensor/control_flow_op.h]: const

File: src/operator/tensor/control_flow_op.h [hunk 4]
```diff
@@ -185,6 +303,62 @@ void WhereOpForward(const nnvm::NodeAttrs& attrs,
...
+template<typename xpu>
+void WhereOpForwardEx(const nnvm::NodeAttrs& attrs,
+                      const OpContext& ctx,
+                      const std::vector<NDArray>& inputs,
+                      const std::vector<OpReqType>& req,
+                      const std::vector<NDArray>& outputs) {
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  const auto& cond_stype = inputs[0].storage_type();
```
Comment (reminisce) [inline src/operator/tensor/control_flow_op.h]: stype is an int, no need to use reference. For simple data types, suggest not use `auto` for better readability.

Comment (eric-haibin-lin) [reply src/operator/tensor/control_flow_op.h]: Good catch!

**Pull request merged**